File: src/637.Average-of-Levels-in-Binary-Tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

// Solution1 -> DFS: Depth First Search
class Solution1 {
public:
    vector<double> averageOfLevels(TreeNode* root) {
        if(!root) return {};
        vector<pair<long long, int>> sum_count;
        vector<double> res;
        // Preorder to fill the sum_count.
        preorder(root, 0, sum_count);
        // Calculate the average number by sum_count.
        for(const auto& p : sum_count)
            res.push_back(static_cast<double>(p.first) / p.second);
        return res;
    }
private:
    void preorder(TreeNode* root, int depth, vector<pair<long long, int>>& sum_count) {
        if(!root) return;
        if(depth >= sum_count.size()) sum_count.push_back({0, 0});
        sum_count[depth].first += root->val;
        sum_count[depth].second += 1;
        preorder(root->left, depth+1, sum_count);
        preorder(root->right, depth+1, sum_count);
    }
};

// Solution2 -> BDS: Breadth First Search
class Solution2 {
public:
    vector<double> averageOfLevels(TreeNode* root) {
        if (root == nullptr) return {};
        // Initialization.
        vector<double> res;
        vector<TreeNode*> curr, next;
        curr.push_back(root);
        // Process every level one by one.
        while (!curr.empty()) {
            long long sum = 0;
            for (const auto& node : curr) {
                sum += node->val;
                if (node->left) next.push_back(node->left);
                if (node->right) next.push_back(node->right);
            }

            res.push_back(static_cast<double>(sum) / curr.size());
            curr.swap(next);
            next.clear();
        }
        return res;
    }
}

File: src/637.Average-of-Levels-in-Binary-Tree/solution.go
package solution

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

// DFS
func averageOfLevels(root *TreeNode) []float64 {
    if root == nil {
        return []float64{}
    }
    sum_count := make([]Counter, 0)
    res := make([]float64, 0)
    // Preorder to fill sum_count
    preorder(root, 0, &sum_count)
    // Calculate average value in each sum_count
    for _, node := range sum_count {
        res = append(res, float64(node.Sum)/float64(node.Num))
    }
    return res

    // // Another way to write this code:
    // // Preorder to fill sum_count
    // sum_count := make([]Counter, 0)
    // preorder(root, 0, &sum_count)
    // // Calculate average value in each sum_count
    // res := make([]float64, len(sum_count))
    // for i, node := range sum_count {
    //     res[i] = float64(node.Sum)/float64(node.Num)
    // }
    // return res
}

type Counter struct {
    Sum int
    Num int
}

func preorder(root *TreeNode, depth int, sum_count *[]Counter) {
    if root == nil {
        return
    }
    if depth >= len(*sum_count) {
        *sum_count = append(*sum_count, Counter{0, 0})
    }
    (*sum_count)[depth].Sum += root.Val
    (*sum_count)[depth].Num += 1
    preorder(root.Left, depth+1, sum_count)
    preorder(root.Right, depth+1, sum_count)
}

// --------------------------------------------------------
// BFS
func averageOfLevels2(root *TreeNode) (averages []float64) {
    nextLevel := []*TreeNode{root}
    for len(nextLevel) > 0 {
        sum := 0
        curLevel := nextLevel
        nextLevel = nil
        for _, node := range curLevel {
            sum += node.Val
            if node.Left != nil {
                nextLevel = append(nextLevel, node.Left)
            }
            if node.Right != nil {
                nextLevel = append(nextLevel, node.Right)
            }
        }
        averages = append(averages, float64(sum)/float64(len(curLevel)))
    }
    return
}
